Setup wizard: Mode selection screen

johnwatkins0 opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses. -->
Fixes #4706

This creates the main framework for the Setup Wizard's mode selection screen. Currently it looks like this:

<img width="1311" alt="Screen Shot 2020-06-18 at 3 38 03 PM" src="https://user-images.githubusercontent.com/9020968/85069443-c1467200-b179-11ea-9f7a-9c4b7da8b13f.png">


This screen serves as a bridge between the theme/plugin scan and the final screens. While it lays out some pieces for those screens, those screens are coming later, so this intermediate step is not fully complete.

### Main changes

- Create the layout and styles for the main components on the screen (shown above)
- Creates the function that will generate all the pieces of copy and the mode recommendations for the 8 different combinations of `hasThemeIssues`, `hasPluginIssues`, and `userIsTechnical`. 
- Adds the React state logic needed to skip the Reader Themes screen if the reader mode is not selected. While working on this I refactored the Reader screen a bit. Previously we were fetching reader theme data at the start of the app in all circumstances. Now we'll only fetch the reader themes when the user has selected reader mode. 

## QA

See QA steps on #4706 

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

johnwatkins0 committed:
Wizard mode select screen

johnwatkins0 committed:
Make screen ui independent of state

johnwatkins0 committed:
Template mode selection screen

johnwatkins0 committed:
CSS updates

johnwatkins0 committed:
Fix failing tests

johnwatkins0 committed:
Move reader theme context back to the top level but only download when mode is set to reader

johnwatkins0 committed:
Change magic numbers to constants

johnwatkins0 committed:
E2E test fix

johnwatkins0 committed:
Fix e2e test

westonruter commented:
While this may come later, I feel like the collapsed sections should have some meaningful information in the collapsed state, namely:

![image](https://user-images.githubusercontent.com/134745/85322336-07058200-b47b-11ea-8267-19a0acf9b273.png)

If all the sections are collapsed, then it will just show Details and Compatibility repeatedly for each mode. I think it would be better to eliminate the “Compatibility” heading entirely, to make that red/green/blue box as a whole an expandable section (moving the arrow inside of it). The duplicated “Not recommended for your site” can then appear inside this section. 

![image](https://user-images.githubusercontent.com/134745/85322539-58ae0c80-b47b-11ea-906e-8c5ee0b68725.png)

The “Details” heading can probably be replaced with a short one-line description of the mode, with then the ability to expand for more information about it. 

cc @amedina @jwold

johnwatkins0 commented:
I like that suggestion. I will give it a try.

amedina commented:
I like that approach as well. The Details and Compatibility headings can be eliminated without loss of information.

jwold commented:
What happens with the one line summary we were going to have?

johnwatkins0 committed:
Template modes design updates

johnwatkins0 committed:
Add learn more links

johnwatkins0 committed:
Update snapshot

johnwatkins0 commented:
I have pushed an update based on the above and the subsequent Slack discussion:

<img width="850" alt="Screen Shot 2020-06-22 at 9 55 11 PM" src="https://user-images.githubusercontent.com/9020968/85355846-1d3f2c80-b4d3-11ea-911b-0f27024dc145.png">

johnwatkins0 committed:
Remove doubled comment

johnwatkins0 committed:
Add missing docblock

johnwatkins0 commented:
@jwold @westonruter We need an interim solution for this for 1.6 since we won't have site scan results. It will mirror what will happen if the site scan fails (which we were planning to account for anyway). Maybe we could use the blue option for every mode _except_ for non-technical users. For non-technical users we can still make reader mode the most recommended option (i.e. the green notice).

We should probably remove the drawer also because we won't have site scan results to display.

jwold commented:
@johnwatkins0 do you mean something like this for the 1.6 release? (cc @amedina and @westonruter)

<img width="1908" alt="PNG image-3B8B64733F53-1" src="https://user-images.githubusercontent.com/3411173/85437968-421db900-b540-11ea-9237-971c286c08b6.png">

The idea being that we have a simple notice at the top for each mode (when relevant), and hold off on the expanding detail box until 1.7.

Then for 1.7 we'd release the other idea we've been discussing: https://d.pr/i/IkQWmC

johnwatkins0 committed:
Handle nonexistent site scan results

johnwatkins0 commented:
@jwold What I have locally right now differs from your sketch in that we still use the large version of the notice component (that will serve as the drawer in 1.7) but it is not expandable, and we show the blue notice for everything. I think we have gone in the direction of the "Details" line being a general explanation of the mode, and the line in the notice being the reason the user may or may not choose that option. In keeping with this, I think it's useful to keep the notice for the "Why?" factor.

**Non-technical users would see:**
<img width="986" alt="Screen Shot 2020-06-23 at 1 10 38 PM" src="https://user-images.githubusercontent.com/9020968/85440067-469f9d00-b553-11ea-95a0-5a29fab4c535.png">

**Technical users would see:**
<img width="924" alt="Screen Shot 2020-06-23 at 1 10 52 PM" src="https://user-images.githubusercontent.com/9020968/85440153-5fa84e00-b553-11ea-9040-0ea6de7dfec5.png">

jwold commented:
@johnwatkins0 thanks for the clarification. I agree with your recommendation! :)

johnwatkins0 committed:
Add e2e tests

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into feature/4706-setup-wizard-mode

johnwatkins0 committed:
Update path to CSS file

pierlon commented:
@johnwatkins0 you can revert https://github.com/ampproject/amp-wp/pull/4900/commits/1bb82541934cdccc90613ccda17a18ae069d6025. That change should not be necessary, as CSS files should not reside in `assets/js`. Along with the revert, removing the optimization config imported from the Gutenberg Webpack config should return the original behavior of the file being put in `assets/css`:

```diff
--- webpack.config.js	(revision 1bb82541934cdccc90613ccda17a18ae069d6025)
+++ webpack.config.js	(date 1592969043556)
@@ -50,7 +50,6 @@
 		} ),
 	],
 	optimization: {
-		...defaultConfig.optimization,
 		minimizer: [
 			new TerserPlugin( {
 				parallel: true,
```

---

As to why it is now putting the `amp-setup` CSS file in `assets/js`:

Due to the `amp-setup` entry being over the default Webpack size limit (244 KiB), Webpack will optimize the asset according to the optimization config supplied, which is currently set to optimize CSS files that match `style.css` and put them in the default output folder (`assets/js`). This configuation was imported from the Gutenberg Webpack config and is not relevant to this project, so it can be removed.

pierlon commented on assets/src/setup/components/amp-notice/index.js:
```diff
@@ -0,0 +1,102 @@
+ * @param {string} props.type The notice type.
+ */
+export function AMPNotice( { children, className, icon, size, type } ) {
+	const noticeIcon = useMemo( () => icon ? icon : getDefaultNoticeIcon( type ), [ icon, type ] );
```
Small nitpick:

Having this all in one line makes this a bit hard to read at first glance. Perhaps having the args on separate lines would help:

```suggestion
	const noticeIcon = useMemo(
		() => icon ? icon : getDefaultNoticeIcon( type ),
		[ icon, type ],
	);
```

pierlon commented on assets/src/setup/components/amp-notice/index.js:
```diff
@@ -0,0 +1,102 @@
+	children: PropTypes.node,
+	className: PropTypes.string,
+	icon: PropTypes.node,
+	size: PropTypes.oneOf( [ NOTICE_SIZE_LARGE, NOTICE_SIZE_SMALL ] ),
```
```suggestion
	size: PropTypes.oneOf( [ NOTICE_SIZE_LARGE, NOTICE_SIZE_SMALL ] ).isRequired,
```

pierlon commented on assets/src/setup/components/amp-notice/index.js:
```diff
@@ -0,0 +1,102 @@
+ *
+ * @param {string} type Notice type.
+ */
+function getDefaultNoticeIcon( type ) {
```
Is there an instance where a custom icon will be needed?

johnwatkins0 replied:
Nah, there was a different icon in one of the earlier comps but it's not there anymore. I'm going to remove the icon prop until it's needed.

pierlon commented on assets/src/setup/pages/template-mode/index.js:
```diff
@@ -1,32 +1,52 @@
 	 */
 	useEffect( () => {
-		if ( canGoForward === false ) {
+		if ( false === canGoForward && false === scanningSite && themeSupport ) {
```
Perhaps we should enforce selecting a technical background before being able to move to the next page.Currently, doing so will result in a intermediate spinner being shown on the next page:

![image](https://user-images.githubusercontent.com/16200219/85501694-67103b80-b5d5-11ea-9ab8-62beea23bc4c.png)

johnwatkins0 replied:
Thanks, that was easy to miss. I've updated.

pierlon commented on assets/src/setup/components/navigation-context-provider.js:
```diff
@@ -41,7 +49,10 @@ export function NavigationContextProvider( { children, pages } ) {
 	const moveForward = () => {
-		setActivePageIndex( activePageIndex + 1 );
+		// Skip the reader theme screen if the user has not selected that mode. Users can go back to it, however.
+		const forwardCount = pages[ activePageIndex + 1 ].slug === 'theme-selection' && 'reader' !== themeSupport ? 2 : 1;
```
A similar thing should happen when navigating backwards. If reader mode is not selected, I can navigate backwards to the "Theme selection" page (although the list of themes is not shown):

![image](https://user-images.githubusercontent.com/16200219/85501833-a3dc3280-b5d5-11ea-8d76-277e7d89700e.png)

pierlon replied:
I think the "Theme selection" page should be removed from the navigation and stepper if reader mode is not selected.

johnwatkins0 replied:
What we decided for this was that we'd skip "Theme Selection" but allow users to return to it. What I'd missed was updating that text to indicate that the screen is not relevant for users who have not selected reader. I've updated.

pierlon commented on assets/src/setup/components/options-context-provider.js:
```diff
@@ -25,7 +25,7 @@ export const Options = createContext();
  */
 export function OptionsContextProvider( { children, optionsRestEndpoint } ) {
-	const [ options, setOptions ] = useState( null );
+	const [ options, setOptions ] = useState( () => {} );
```
What's the difference of simply setting the state to `{}`?

johnwatkins0 replied:
Yeah, that wasn't necessary.

pierlon commented on assets/src/setup/components/options-context-provider.js:
```diff
@@ -126,5 +130,5 @@ export function OptionsContextProvider( { children, optionsRestEndpoint } ) {
 OptionsContextProvider.propTypes = {
 	children: PropTypes.any,
-	optionsRestEndpoint: PropTypes.string.isRequired,
+	optionsRestEndpoint: PropTypes.string,
```
Any reason why this is no longer required?

johnwatkins0 replied:
This was for unit testing purposes, but I've revised the approach; rather than changing the prop here (it should be required), in the test I've just mocked an OptionsContextProvider that doesn't fetch.

pierlon commented on assets/src/setup/components/site-scan-context-provider.js:
```diff
@@ -0,0 +1,69 @@
+	const [ scanningSite, setScanningSite ] = useState( true );
+
+	/**
+	 * @todo This temporary code is for development purposes.
```
Are these still needed for development purposes? I suppose they can be removed before merge, or simply replaced when the site scanning feature is ready.

johnwatkins0 replied:
I've updated this comment to explain more what's going on. Essentially we need to keep everything here, but right now this component is passing `null` for theme issues and plugin issues, imitating what will happen in the future if the site scan is unsuccessful for some reason.

pierlon commented on assets/src/setup/pages/template-mode/get-selection-details.js:
```diff
@@ -0,0 +1,206 @@
+ * @param {Object} args
+ * @param {boolean} args.userIsTechnical Whether the user answered yes to the technical question.
+ * @param {boolean} args.hasPluginIssues Whether the site scan found plugin issues.
+ * @param {boolean} args.hasThemeIssues Whether the site scan found theme issues.
```
Missing param doc for `hasScanResults`.

pierlon commented on assets/src/setup/pages/template-mode/get-selection-details.js:
```diff
@@ -0,0 +1,206 @@
+			return 'Standard mode is a good choice if your site has an AMP-compatible theme and only uses AMP-compatible plugins. If you\'re not sure of the compatibility of your themes and plugins, Reader mode may be a better option.';
+
+		default: {
+			throw new Error( __( 'A selection text recommentation was not accounted for. ', 'amp' ) + JSON.stringify( args ) );
```
```suggestion
			throw new Error( __( 'A selection text recommendation was not accounted for. ', 'amp' ) + JSON.stringify( args ) );
```

pierlon commented on assets/src/setup/pages/template-mode/index.js:
```diff
@@ -1,32 +1,52 @@
+		return <Loading />;
+	}
 
+	// The actual display component should avoid using global context directly. This will facilitate developing and testing the UI using different options..
```
```suggestion
	// The actual display component should avoid using global context directly. This will facilitate developing and testing the UI using different options.
```

pierlon commented on assets/src/setup/pages/template-mode/screen-ui.js:
```diff
@@ -0,0 +1,192 @@
+ * @param {boolean} props.selected Whether the mode is selected.
+ * @param {string} props.title The title for the selection.
+ */
+export function Selection( { compatibility, id, illustration, details, onChange, recommended, selected, title } ) {
```
Missing `@param` for `id`.

pierlon commented on assets/src/setup/pages/template-mode/test/get-selection-details.js:
```diff
@@ -0,0 +1,32 @@
+/**
+ * Internal dependencies
+ */
+import { getRecommendationLevels, getSelectionText, COMPATIBILITY, DETAILS, READER, STANDARD, TRANSITIONAL, MOST_RECOMMENDED, NOT_RECOMMENDED, RECOMMENDED, TECHNICAL, NON_TECHNICAL } from '../get-selection-details';
```
Maybe put each import on a new line?

pierlon commented on tests/e2e/specs/amp-onboarding/index.js:
```diff
@@ -16,9 +16,14 @@ describe( 'AMP Setup Screen', () => {
-	const getTitleText = () => page.$eval( 'h1', ( el ) => el.innerText );
+	const getTitleText = async () => {
+		await page.waitForSelector( 'h1' );
+		const titleText = await page.$eval( 'h1', ( el ) => el.innerText );
+		return titleText;
```
```suggestion
		return page.$eval( 'h1', ( el ) => el.innerText );
```

johnwatkins0 committed:
Revert change to CSS file placement

johnwatkins0 committed:
Remove unused AMPNotice icon prop

johnwatkins0 committed:
Merge changes from code review

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

johnwatkins0 committed:
Update theme selection text when reader mode is not selected

johnwatkins0 committed:
Remove unnecessary state initialization

johnwatkins0 committed:
Remove unnecessary state initialization

johnwatkins0 committed:
Lint fixes

johnwatkins0 committed:
Update comment explaining temporary code

johnwatkins0 committed:
Require technical background selection before moving forward

johnwatkins0 committed:
Break long import into separate lines

johnwatkins0 committed:
Re-require options context rest endpoint prop

johnwatkins0 committed:
Fix reader themes e2e test

johnwatkins0 committed:
Nav e2e test fix

johnwatkins0 committed:
Additional check in e2e beforEach function

johnwatkins0 commented:
Thanks for the review @pierlon . I have addressed all the feedback. @westonruter Let me know if you see anything else, noting the update I've made to handle the absence of site scan results in 1.6 (see screenshot above).

westonruter commented on assets/src/setup/pages/choose-reader-theme/index.js:
```diff
@@ -21,25 +21,38 @@ export function ChooseReaderTheme() {
+	if ( 'reader' !== themeSupport ) {
+		return (
+			<p>
+				{ __( 'This screen is only relevant to sites that use Reader mode. Please continue', 'amp' ) }
```
```suggestion
				{ __( 'This screen is only relevant to sites that use Reader mode. Please continue.', 'amp' ) }
```

westonruter commented on assets/src/setup/pages/template-mode/screen-ui.js:
```diff
@@ -0,0 +1,193 @@
+					<span dangerouslySetInnerHTML={ { __html: details } } />
+					{ ' ' }
+					{ /* @todo Temporary URL. */ }
+					<a href="http://amp-wp.org">
```
```suggestion
					<a href="http://amp-wp.org" target="_blank">
```

westonruter commented on assets/src/setup/pages/template-mode/screen-ui.js:
```diff
@@ -0,0 +1,193 @@
+						<p>
+							{ 'Lorem ipsum dolar sit amet. ' }
+							{ /* @todo Temporary URL. */ }
+							<a href="http://amp-wp.org">
```
```suggestion
							<a href="http://amp-wp.org" target="_blank">
```

westonruter reviewed (approved):
Good evolution on the screens.
